### ISSUE TITLE
Log Entries created on remote/other Peers are not Pinned Locally

### ISSUE THREAD

**Issue by justin0mcateer** [orbitdb/orbitdb#1163]
In the previous implementation of OpLog.joinEntry, Log entries retrieved from other nodes were 'put' in the IPFSBlockStorage and optionally (via configuration) pinned. In the rewrite of OpLog.joinEntry _entries.put is no longer called and so the remotely created Log entries are no longer stored locally.

This results in the behavior that _all_ Peers which have previously written to the log MUST be online or traversal of the Log will fail and unhandled exception in IPFSBlockStorage.get. This does not seem to be desirable behavior.

This code will crash with an unhandled promise rejection if any block fails to be fetched, which currently appears to occur if any previous log writer is offline:
https://github.com/orbitdb/orbitdb/blob/main/src/storage/ipfs-block.js#L61

In older versions of the code (prior to the OpLog.joinEntry rewrite), fetched OpLog entries were persisted locally:
https://github.com/orbitdb/orbitdb/blob/v1.0.0/src/oplog/log.js#L269

**Comment by haydenyoung** [comment]
Is the peer using an in-memory blockstore for Helia? What happens if you configure the node with a persistent blockstore?

E.g.

```
import { LevelBlockstore } from 'blockstore-level' 

const blockstore = new LevelBlockstore('./some/path')
const libp2p = await createLibp2p()
const helia = await createHelia({ libp2p, blockstore })
```

My understanding is that the replicated block should also be pinned on the remote peer (`isPinned` should == "true" on the remote peer also).

**Comment by haydenyoung** [comment]
Just to add that a persistent datastore may also be needed as I believe pins are stored to datastore not blockstore.

```
import { LevelBlockstore } from 'blockstore-level' 
import { LevelDatastore } from 'datastore-level' 

const blockstore = new LevelBlockstore('./some/path/helia/blocks')
const datastore = new LevelDatastore('./some/path/heliadata')
const libp2p = await createLibp2p()
const helia = await createHelia({ libp2p, blockstore, datastore })
```

(sorry haven't been able to actually confirm this, just going on a hunch re: datastore and pins).

**Comment by justin0mcateer** [comment]
We are using a persistent store (IDB) for both. What I am saying is that the new logic doesn't call Helia.put for remote originated blocks at all, at least from what I can see.

There was an explicit call to _entries.put prior to the joinEntry rewrite. I've looked and I can't see it being done anywhere else. We are seeing block not found errors when not all the nodes are online.


________________________________

**Comment by justin0mcateer** [comment]
>My understanding is that the replicated block should also be pinned on the remote peer (isPinned should == "true" on the remote peer also).

I am not really sure what you are referring to here. The only way to pin anything in Helia is either locally, or via the 'Pinning API'. There isn't any way to compel another node to Pin your data.

**Comment by haydenyoung** [comment]
Does IPFSBlockStorage fail immediately (I.e. as soon as the remote node is offline) or does it happen over time (I.e. the remote node has been offline for some time)? What would be the steps for me to reproduce this issue?

**Comment by ikoenigsknecht** [comment]
Is there any update on this?  I am seeing the same issue in my project.  When testing with two peers and a ton of logging manually added to my dependencies I see that when peer A receives a message from peer B it gets the data and thus the message can be displayed but if you close out of both peers and reopen peer A it will crash on a `Want for <cid> aborted` error because the blockstore was never written to.  If, however, peer A writes a message to the DB on top of the one it received from peer B it will pull the blocks it was missing and then peer B becomes the peer with missing blocks.  

I tested by adding back the `await _entries.put(entry.hash, entry.bytes)` line to `log.joinEntry` and that seems to fix things, at least in my case, and I see the blocks get written when created by the other peer.

My questions are:
1. Is there a particular reason this was removed as part of the 2.0.0 upgrade?
2. Is there any objection to adding that back?

**Comment by ikoenigsknecht** [comment]
> We've been running a patch that re-added the _entries.put since opened this issue and we haven't had any more issues with that.

That was my plan, as well.

**Comment by haadcode** [comment]
`entries.put()` shouldn’t be needed as `get()` is called when fetching the entries to join, so the blocks should be in blockstore of all peers who have replicated the db.

Would you be able to set up a test where the reported behavior can be observed?

**Comment by justin0mcateer** [comment]
The problem is that 'put' will pin the blocks; 'get' will fetch the blocks locally, but it will not pin the blocks. Hence, the next time the garbage collection runs, the blocks will be lost locally. IFF, another node which has the blocks pinned happens to be online at the time the blocks are needed, everything will work because the block will be (needlessly) fetched from the network again. However, if the block is garbage collected and another peer with the block is not online, the block will be unavailable and will fail to be fetched. This failure to fetch will immediately lead to failure to traverse the log.

Replicating some data between two peers each providing log entries, taking one node offline, running garbage collection and restarting the remaining node should be sufficient to demonstrate the problem.
________________________________

**Comment by haadcode** [comment]
Ah, so this is related to garbage collection, it sounds. Where can we read about how garbage collection works in Helia? How does one trigger it manually?

Regardless, it sounds like what we’re missing is “pin on receive” and for that I suppose it makes sense to add back `_entries.put(…)` as suggested.

Would either of you want to make a PR for your patch that addresses this?

**Comment by abuvanth** [comment]
> Ah, so this is related to garbage collection, it sounds. Where can we read about how garbage collection works in Helia? How does one trigger it manually?
> 
> Regardless, it sounds like what we’re missing is “pin on receive” and for that I suppose it makes sense to add back `_entries.put(…)` as suggested.
> 
> Would either of you want to make a PR for your patch that addresses this?

yes bring `_entires.put(entry.hash, entry.bytes)` back to joinEntry should fix this issue.

**Comment by cristianmadularu** [comment]
Any udpate for this? This is a serious problem for us. Thank you!

**Comment by abuvanth** [comment]
not this point. 

if db.close() called then db will not be sync

**Comment by haydenyoung** [comment]
https://github.com/orbitdb/orbitdb/pull/1198

**Comment by haydenyoung** [comment]
Fix in release [v2.3.0](https://github.com/orbitdb/orbitdb/tree/v2.3.0).
